Improve the Structure of the OpenAPI Guide (#4150)
ballerina-platform/ballerina-dev-website: learn/guides/ballerina-openapi-support.md

  - /learn/cli-documentation/openapi
---

The Ballerina OpenAPI tooling support provides the following capabilities.

> **Prerequisite**: Download and install the Ballerina Swan Lake from [downloads](https://ballerina.io/downloads/).

 1. Generating Ballerina service/client stubs from a given OpenAPI contract file using the CLI tool.
 2. Exporting the OpenAPI definition from a given Ballerina service implementation using the CLI tool.
 3. Validating the service implementation compliance with a provided OpenAPI contract using the OpenAPI annotation.
    - The OpenAPI compiler plugin will allow you to validate a service implementation against an OpenAPI contract during
  the compile time. This plugin ensures that the implementation of a service does not deviate from its OpenAPI contract. 

 > **Note**: Ballerina Swan Lake supports the OpenAPI Specification version 3.0.0 onwards.

## Generate a Ballerina service 

> Generating a Ballerina service from an OpenAPI definition is named as the `design-first approach`. To generate only the Ballerina service, you can use the following CLI command of the OpenAPI tool.

Reviewer: Following the  `API design-first approach`,  you can generate only the Ballerina service from an OpenAPI definition, using the below CLI command of the OpenAPI tool.

Reviewer: ```suggestion
Following the  `API design-first approach`,  you can generate only the Ballerina service from an OpenAPI definition, using the below CLI command of the OpenAPI tool.
```

-- hello_service.bal
-- types.bal
```
### Generate from tags
To generate the Ballerina service stub with a subset of tags defined in an OpenAPI contract, use the `--tags` option and specify the tags you need as specified in the OpenAPI definition.
```bash
$ bal openapi -i <openapi-contract> [--tags <"tag1","tag2">]
```

For example, 

```bash
$ bal openapi -i hello.yaml --tags "pets", "list"
```
Once you execute the command, only the operations related to the given tags will be included in the generated service file.

>**Info:** For more command options, see [Command reference](#command-reference).

## Export an OpenAPI contract

> Exporting an OpenAPI contract from a Ballerina service is named as the `code-first approach`. You can convert your Ballerina service APIs into human-readable or machine-readable documents such as OpenAPI documents by using the Ballerina to OpenAPI CLI Tool as follows.

Reviewer: Following the `API code-first approach`, you can convert your Ballerina service APIs into human-readable or machine-readable documents such as OpenAPI documents by using the Ballerina to OpenAPI CLI Tool as follows.

Reviewer: ```suggestion
Following the `API code-first approach`, you can convert your Ballerina service APIs into human-readable or machine-readable documents such as OpenAPI documents by using the Ballerina to OpenAPI CLI Tool as follows.
```
